### PR TITLE
Use Path.arc() to interpolate polar arcs.

### PR DESCRIPTION
**Pull request by anntzer**
Currently, in polar plots, arcs are rendered by linearly interpolating
in (theta, r) space before conversion to (x, y) space (if their
_interpolation_steps attribute is >1 -- this attribute is described in
the docs as "an implementation detail not intended for public use").

When _interpolation_steps == 1, this PR doesn't change anything -- there
is an example (specialty_plots/radar_chart) that explicitly relies on
being able to set _interpolation_steps to 1 and get a "polygonal" polar
axes rather than a circular one.

When _interpolation_steps > 1, however, this PR looks for paths segments
that are either at a constant theta or a constant r.  It transforms
constant-theta segments to a single segment (because interpolation
doesn't help there) and transforms constant-r segments (circular arcs)
to a Bezier approximation of a circle which is already implemented in
Path.arc()).

This greatly decreases the number of control points for such plots in
vector outputs (which are thus smaller -- note the line counts changes
in the svg files...), and also improves rasterization by mplcairo (cairo
appears to be not so good at filling areas with a lot of close vertices).

The many changes in baseline images are due to, well, the change of
approximation for drawing circles.  (Note that if we agree with the
approach of this PR, I could perhaps do a first PR to decrease the
number of polar tests so that I don't have to commit so many new
images...)

Example: `polar(); bar([0], [0.9])`

matplotlib svg before, showing control points in inkscape:
oldsvg

matplotlib svg after, showing control points in inkscape:
newsvg

mplcairo before:
old

mplcairo after:
new

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
This looks really useful.  Can the old images be kept by setting interpolation_steps=1?  OTOH I think if a default changes we should consider buying the bullet and changing the test images.  Same for the image antialiasing PR

**Comment by anntzer** [comment]
No, interpolation_steps=1 results in something like https://matplotlib.org/3.1.0/gallery/specialty_plots/radar_chart.html.

I'm fine with modifying a few baseline images but I think it's also a good opportunity to see whether we can combine some tests...

File: lib/matplotlib/projections/polar.py [hunk 1]
```diff
@@ -57,11 +57,57 @@ def transform_non_affine(self, tr):
...
+                    # The following is complicated by Path.arc() being
+                    # "helpful" and unwrapping the angles, but we don't want
+                    # that behavior here.
+                    last_td, td = np.rad2deg([last_t, t])
+                    if self._use_rmin and self._axis is not None:
+                        r = ((r - self._axis.get_rorigin())
+                             * self._axis.get_rsign())
+                    if last_td <= td:
+                        while td - last_td > 360:
+                            arc = Path.arc(last_td, last_td + 360)
```
Comment (QuLogic) [inline lib/matplotlib/projections/polar.py]: This block seems untested.

File: lib/matplotlib/projections/polar.py [hunk 2]
```diff
@@ -57,11 +57,57 @@ def transform_non_affine(self, tr):
...
+                            codes.extend(arc.codes[1:])
+                            last_td += 360
+                        arc = Path.arc(last_td, td)
+                        xys.extend(arc.vertices[1:] * r)
+                        codes.extend(arc.codes[1:])
+                    else:
+                        # The reverse version also relies on the fact that all
+                        # codes but the first one are the same.
+                        while last_td - td > 360:
+                            arc = Path.arc(last_td - 360, last_td)
```
Comment (QuLogic) [inline lib/matplotlib/projections/polar.py]: As does this block.

File: lib/matplotlib/projections/polar.py [hunk 3]
```diff
@@ -57,11 +57,57 @@ def transform_non_affine(self, tr):
...
+                            xys.extend(arc.vertices[::-1][1:] * r)
+                            codes.extend(arc.codes[1:])
+                            last_td -= 360
+                        arc = Path.arc(td, last_td)
+                        xys.extend(arc.vertices[::-1][1:] * r)
+                        codes.extend(arc.codes[1:])
+                else:  # Interpolate.
+                    trs = cbook.simple_linear_interpolation(
```
Comment (QuLogic) [inline lib/matplotlib/projections/polar.py]: And this block too.

File: lib/matplotlib/projections/polar.py [hunk 4]
```diff
@@ -57,11 +57,57 @@ def transform_non_affine(self, tr):
...
+                        np.column_stack([(last_t, last_r), trs]),
+                        path._interpolation_steps)[1:]
+                    xys.extend(self.transform_non_affine(trs))
+                    codes.extend(Path.LINETO for _ in trs)
```
Comment (QuLogic) [inline lib/matplotlib/projections/polar.py]: Is this really better than `codes.extend([Path.LINETO] * len(trs))`?

Comment (anntzer) [reply lib/matplotlib/projections/polar.py]: either way works for me, switched to your preference.

**Comment by anntzer** [comment]
Rebased on top of #14888 which deletes a few baseline images (replacing them by check_figures_equal).

**Comment by anntzer** [comment]
Also, note that with this patch, one cannot anymore set e.g. GRIDLINE_INTERPOLATION_STEPS to a small but >1 number (e.g. 3) and see 3-point interpolation on circular arcs, but I'm not too worried about that...

File: lib/matplotlib/projections/polar.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/projections/polar.py]: test added, although it's not particularly interesting (I can't actually think of a very interesting test for that...)

File: lib/matplotlib/projections/polar.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/projections/polar.py]: same as above

File: lib/matplotlib/projections/polar.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/projections/polar.py]: test added

**Comment by timhoffm** [comment]
I'm positive on the idea of this change. Ping when you think it's ready for review.

**Comment by anntzer** [comment]
This PR includes #14888 so it can already be reviewed (at worst this one gets merged before #14888 and both commits go in at the same time).  (I don't think the other polar baseline images can "easily" be removed.)

**Comment by QuLogic** [comment]
Docs fail, might need a rebase.

**Comment by anntzer** [comment]
rebased

**Comment by anntzer** [comment]
I don't mind waiting to see at least whether the test-images-infrastructure-rewrite GSoC will happen or not (after all this has already been there for 8 months, an extra couple of them isn't too bad...).

**Comment by tacaswell** [comment]
Decided to merge as-is as this is a pretty big improvement for SVG and mplcairo.

**Pull request merged**